File: src/types/source.ts
export interface Source {
  id: string
  subjectId: string | null
  book: string | null
  chapter: string | null
  knowledge: string | null
  createdAt: string
  updatedAt: string
}

export interface CreateSourceRequest {
  subjectId?: string | null
  book?: string | null
  chapter?: string | null
  knowledge?: string | null
}

export interface CreateSourceResponse {
  source: Source
}

export interface UpdateSourceRequest {
  id: string
  subjectId: string | null
  book: string | null
  chapter: string | null
  knowledge: string | null
}

export interface UpdateSourceResponse {
  source: Source
}

export interface DeleteSourceRequest {
  id: string
}

export interface DeleteSourceResponse {
  id: string
}

export interface DeleteSourcesRequest {
  ids: string[]
}

export interface DeleteSourcesResponse {
  ids: string[]
}

export interface GetSourceRequest {
  id: string
}

export interface GetSourceResponse {
  source: Source
}

export interface ListSourcesRequest {
  subjectId?: string | null
}

export interface ListSourcesResponse {
  sources: Source[]
}

File: src-tauri/src/command/response/source.rs
use serde::Serialize;

use crate::domain::model::Source;

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct SourceData {
    pub(crate) id: String,
    pub(crate) subject_id: Option<String>,
    pub(crate) book: Option<String>,
    pub(crate) chapter: Option<String>,
    pub(crate) knowledge: Option<String>,
    pub(crate) created_at: String,
    pub(crate) updated_at: String,
}

impl From<Source> for SourceData {
    fn from(value: Source) -> Self {
        Self {
            id: value.id.to_string(),
            subject_id: value.subject_id.map(|id| id.to_string()),
            book: value.book,
            chapter: value.chapter,
            knowledge: value.knowledge,
            created_at: value.metadata.created_at.to_rfc3339(),
            updated_at: value.metadata.updated_at.to_rfc3339(),
        }
    }
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateSourceResponse {
    pub(crate) source: SourceData,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct UpdateSourceResponse {
    pub(crate) source: SourceData,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DeleteSourceResponse {
    pub(crate) id: String,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DeleteSourcesResponse {
    pub(crate) ids: Vec<String>,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct GetSourceResponse {
    pub(crate) source: SourceData,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ListSourcesResponse {
    pub(crate) sources: Vec<SourceData>,
}
